zkevm-circuits/src/evm_circuit/execution/log.rs

```diff
@@ -0,0 +1,630 @@
+        let is_persistent = cb.call_context(None, CallContextFieldTag::IsPersistent);
+        cb.condition(is_persistent.expr(), |cb| {
+            cb.tx_log_lookup(
+                1.expr(),
+                tx_id.expr(),
+                TxLogFieldTag::Address,
+                0.expr(),
+                contract_address.expr(),
+            );
+        });
+
+        // constrain topics in logs
+        let topics = array_init(|_| cb.query_cell());
+        let is_topic_zeros: [IsZeroGadget<F>; 4] =
+            array_init(|idx| IsZeroGadget::construct(cb, topics[idx].expr()));
+        for (idx, topic) in topics.iter().enumerate() {
+            cb.condition(1.expr() - is_topic_zeros[idx].expr(), |cb| {
+                cb.stack_pop(topic.expr());
+                //cb.tx_log_lookup(1.expr(), tx_id.expr(),
+                // TxLogFieldTag::Topic, idx.expr(), topic.expr());
+            });
+            cb.condition(
+                (1.expr() - is_topic_zeros[idx].expr()) * is_persistent.expr(),
+                |cb| {
+                    cb.tx_log_lookup(
+                        1.expr(),
+                        tx_id.expr(),
+                        TxLogFieldTag::Topic,
+                        idx.expr(),
+                        topic.expr(),
+                    );
+                },
+            );
+        }
+
+        let opcode = cb.query_cell();
+        let topic_count = opcode.expr() - OpcodeId::LOG0.as_u8().expr();
+        let selector_exprs = is_topic_zeros.clone().map(|idx| idx.expr());
+
+        // TOPIC_COUNT == Non zero topic count
+        cb.require_equal(
+            " sum of topic selectors = topic_count ",
+            4.expr() - topic_count.clone(),
+            sum::expr(selector_exprs),
+        );
+        // `is_topic_zeros` order must be from 0 --> 1
+        for idx in 1..4 {
+            let selector_prev = is_topic_zeros[idx - 1].expr();
+            // selector can transit from 0 to 1 only once as [0, ..., 1]
+            cb.require_boolean(
+                "Constrain topic selectors can only transit from 0 to 1",
+                is_topic_zeros[idx].expr() - selector_prev,
+            );
+        }
+
+        // check memory copy
+        let memory_address =
+            MemoryAddressGadget::construct(cb, memory_offset.clone(), msize.clone());
+
+        // Calculate the next memory size and the gas cost for this memory
+        // access
+        let memory_expansion = MemoryExpansionGadget::construct(
+            cb,
+            cb.curr.state.memory_word_size.expr(),
+            [memory_address.address()],
+        );
+
+        // If the iterative process has not yet finished, we constrain the next step to
+        // be another `CopyToLog` while adding some additional
+        // constraints to the auxiliary data.
+        // Constrain the next step CopyToLog if length != 0
+        cb.constrain_next_step(
```

Comment:
if it's not persistent, do we still need to perform the copy?

Reply:
That's a good point!  The spec is currently defined like this (when it's not persistent, there's still the copy steps but they do nothing).
In case we want to change this behavior, we should first update the spec.

My personal opinion is that we can leave it like this for this PR, and if we decide to change it do it via another PR in the future.

Reply:
still need to do the memory reading,  if not persistent, no log added to state .